### ISSUE TITLE
Can ELK integrate with Spring ?

### ISSUE THREAD

**Issue by xutong8** [eclipse/elk#654]
```java
@Controller
public class HelloWorldController {
	
	@Autowired
	private AbstractLayoutProvider provider;
	
	@RequestMapping(value = "/hello", method = RequestMethod.GET)
	@ResponseBody
	public String hello() {

		String graphStr = "{\"id\":\"root\",\"children\":[{\"id\":\"n1\",\"width\":30,\"height\":30},{\"id\":\"n2\",\"width\":30,\"height\":30},{\"id\":\"n3\",\"width\":30,\"height\":30}],\"edges\":[{\"id\":\"e1\",\"sources\":[\"n1\"],\"targets\":[\"n2\"]},{\"id\":\"e2\",\"sources\":[\"n1\"],\"targets\":[\"n3\"]}]}";

		ElkNode node = ElkGraphJson.forGraph(graphStr).toElk();
		node.setProperty(CoreOptions.ALGORITHM, "org.eclipse.elk.layered");
		System.out.println(!node.hasProperty(CoreOptions.RESOLVED_ALGORITHM));
		System.out.println(!node.getChildren().isEmpty());
		
		RecursiveGraphLayoutEngine engine = new RecursiveGraphLayoutEngine();
		BasicProgressMonitor progressMonitor = new BasicProgressMonitor();
		engine.layout(node, progressMonitor);
		
		return "Hello World";
	}
}
```

My code was described above. I attempted to make ELK integrate with spring. But i encountered a problem that was about "org.eclipse.elk.core.UnsupportedConfigurationException: Layout algorithm 'org.eclipse.elk.layered' not found for Root Node root".

**Comment by xutong8** [comment]
Excuse me, Is that my problem? @uruuru

**Comment by soerendomroes** [comment]
Do you have access to `LayeredOptions.ALGORITHM_ID`?

**Comment by xutong8** [comment]
No, I don't access to LayeredOptions.ALGORITHM_ID.

`
org.eclipse.elk.core.UnsupportedConfigurationException: Layout algorithm 'org.eclipse.elk.layered' not found for Root Node root
	at org.eclipse.elk.core.data.LayoutAlgorithmResolver.resolveAlgorithm(LayoutAlgorithmResolver.java:53) ~[org.eclipse.elk.core-0.6.1.jar:na]
	at org.eclipse.elk.core.data.LayoutAlgorithmResolver.visit(LayoutAlgorithmResolver.java:28) ~[org.eclipse.elk.core-0.6.1.jar:na]
	at org.eclipse.elk.core.util.ElkUtil.applyVisitors(ElkUtil.java:873) ~[org.eclipse.elk.core-0.6.1.jar:na]
	at org.eclipse.elk.core.RecursiveGraphLayoutEngine.layout(RecursiveGraphLayoutEngine.java:89) ~[org.eclipse.elk.core-0.6.1.jar:na]
	at org.eclipse.elk.core.RecursiveGraphLayoutEngine.layout(RecursiveGraphLayoutEngine.java:71) ~[org.eclipse.elk.core-0.6.1.jar:na]
	at com.zju.elkgraph.controller.HelloWorldController.hello(HelloWorldController.java:33) ~[classes/:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_161]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_161]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_161]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_161]
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:190) ~[spring-web-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:138) ~[spring-web-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:106) ~[spring-webmvc-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:879) ~[spring-webmvc-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:793) ~[spring-webmvc-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87) ~[spring-webmvc-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1040) ~[spring-webmvc-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:943) ~[spring-webmvc-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006) ~[spring-webmvc-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:898) ~[spring-webmvc-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:634) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883) ~[spring-webmvc-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:741) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53) ~[tomcat-embed-websocket-9.0.31.jar:9.0.31]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100) ~[spring-web-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119) ~[spring-web-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93) ~[spring-web-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119) ~[spring-web-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201) ~[spring-web-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119) ~[spring-web-5.2.4.RELEASE.jar:5.2.4.RELEASE]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202) ~[tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96) [tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:541) [tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139) [tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92) [tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74) [tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343) [tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:367) [tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65) [tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:868) [tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1639) [tomcat-embed-core-9.0.31.jar:9.0.31]
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) [tomcat-embed-core-9.0.31.jar:9.0.31]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_161]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_161]
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) [tomcat-embed-core-9.0.31.jar:9.0.31]
	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_161]
`

**Comment by soerendomroes** [comment]
The `layered` algorithm is in the `org.eclipse.elk.alg.layered` bundle. If you cannot access `LayeredOptions` this bundle my be missing.

**Comment by xutong8** [comment]
yeah. I just added this bundle. But a new problem took place. I never saw it from the official document.
` java.lang.IllegalStateException: Couldn't clone property 'org.eclipse.elk.nodeLabels.padding'. Make sure it's type is registered with the ElkReflect utility class.
	at org.eclipse.elk.graph.properties.Property.getDefault(Property.java:137) ~[org.eclipse.elk.graph-0.6.0.jar:na]
	at org.eclipse.elk.core.util.IndividualSpacings.getIndividualOrInherited(IndividualSpacings.java:86) ~[org.eclipse.elk.core-0.6.1.jar:na]
	at org.eclipse.elk.alg.common.nodespacing.internal.NodeContext.<init>(NodeContext.java:140) ~[org.eclipse.elk.alg.common-0.6.0.jar:na]
	at org.eclipse.elk.alg.common.nodespacing.NodeLabelAndSizeCalculator.computeInsideNodeLabelPadding(NodeLabelAndSizeCalculator.java:209) ~[org.eclipse.elk.alg.common-0.6.0.jar:na]
	at org.eclipse.elk.alg.layered.graph.transform.ElkGraphImporter.createLGraph(ElkGraphImporter.java:457) ~[org.eclipse.elk.alg.layered-0.6.1.jar:na]
	at org.eclipse.elk.alg.layered.graph.transform.ElkGraphImporter.importGraph(ElkGraphImporter.java:86) ~[org.eclipse.elk.alg.layered-0.6.1.jar:na]
	at org.eclipse.elk.alg.layered.graph.transform.ElkGraphTransformer.importGraph(ElkGraphTransformer.java:37) ~[org.eclipse.elk.alg.layered-0.6.1.jar:na]
	at org.eclipse.elk.alg.layered.graph.transform.ElkGraphTransformer.importGraph(ElkGraphTransformer.java:1) ~[org.eclipse.elk.alg.layered-0.6.1.jar:na]
	at org.eclipse.elk.alg.layered.LayeredLayoutProvider.layout(LayeredLayoutProvider.java:48) ~[org.eclipse.elk.alg.layered-0.6.1.jar:na]
	at com.zju.elkgraph.controller.HelloWorldController.hello(HelloWorldController.java:29) ~[classes/:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_161]`

**Comment by uruuru** [comment]
You have to call `PlainJavaInitialization.initializePlainJavaLayout();` before calling the layout. 

@le-cds, @soerendomroes I guess we should add this call to the following page, shouldn't we? https://www.eclipse.org/elk/documentation/tooldevelopers/usingplainjavalayout.html

**Comment by le-cds** [comment]
@uruuru That's an unfortunate result of the fact that the website is built from the master branch... IIRC, the information were there until I implemented service loaders as part of #402, at which point I changed the website. This should work as is as soon as 0.7.0 is released.

**Comment by uruuru** [comment]
> IIRC, the information were there until I implemented service loaders 

That explains why I was expecting the information to be on the website :D

**Comment by uruuru** [comment]
I just realized that the `PlainJavaInitialization` is part of the test plugins. 

@xutong8 please refer to the following state of the documentation on plain java layout for 0.6.1 and before: 
https://github.com/eclipse/elk/blob/22c8f7bc05bb152f9e52f9d3d82036f8d3b60b2d/docs/content/documentation/tooldevelopers/usingplainjavalayout.md

TLDR;
```
LayoutMetaDataService service = LayoutMetaDataService.getInstance();
service.registerLayoutMetaDataProviders(new LayeredMetaDataProvider());
```

**Comment by le-cds** [comment]
> That explains why I was expecting the information to be on the website :D

Yep... #291 would still be a welcome addition...

**Comment by xutong8** [comment]
> I just realized that the `PlainJavaInitialization` is part of the test plugins.
> 
> @xutong8 please refer to the following state of the documentation on plain java layout for 0.6.1 and before:
> https://github.com/eclipse/elk/blob/22c8f7bc05bb152f9e52f9d3d82036f8d3b60b2d/docs/content/documentation/tooldevelopers/usingplainjavalayout.md
> 
> TLDR;
> 
> ```
> LayoutMetaDataService service = LayoutMetaDataService.getInstance();
> service.registerLayoutMetaDataProviders(new LayeredMetaDataProvider());
> ```

Thank you very much！I solved the problem.
